overlord/registrystate/registrystate.go

```diff
@@ -197,45 +198,120 @@ var writeDatabag = func(st *state.State, databag registry.JSONDataBag, account,
 	return nil
 }
 
-type cachedRegistryTx struct {
-	account  string
-	registry string
-}
+// GetTransaction retrieves or creates a transaction to access the view's
+// registry. The state must be locked by the caller.
+func GetTransaction(ctx *Context, st *state.State, view *registry.View) (*Transaction, error) {
+	account, registryName := view.Registry().Account, view.Registry().Name
+
+	// check if we're already running in the context of a committing transaction
+	hookCtx := ctx.hookCtx
+	if hookCtx != nil && !hookCtx.IsEphemeral() && IsRegistryHook(hookCtx) {
+		// running in the context of a transaction, so if the referenced registry
+		// doesn't match that tx, we only allow the caller to read the other registry
+		t, _ := hookCtx.Task()
+		tx, commitTask, err := GetStoredTransaction(t)
+		if err != nil {
+			return nil, fmt.Errorf("cannot access registry view %s/%s/%s: cannot get transaction: %v", account, registryName, view.Name, err)
+		}
+
+		if tx.RegistryAccount != account || tx.RegistryName != registryName {
+			// TODO: can we allow accessing a different registry just for reading? we'll
+			// need to create a change and run hooks so will require accounting in state
+			return nil, fmt.Errorf("cannot access registry %s/%s: ongoing transaction for %s/%s", account, registryName, tx.RegistryAccount, tx.RegistryName)
+		}
+
+		ctx.onDone(func() error {
+			setTransaction(commitTask, tx)
+			return nil
+		})
```

Comment:
this bit merits a comment probably, if I understand this will run at the end of the successful hook? also is setTransaction idempotent if the transaction existed already?

Reply:
I'm not actually completely sure why we wait the end of the hook to do this?

Reply:
setTransaction is idempotent and the transaction is expected to exist, this is used to update it in the state. We could do it after each command but I thought this was cleaner and more efficient (since a hook may do many snapcl sets)